sympy/physics/mechanics/functions.py

```diff
@@ -532,7 +539,7 @@ def _recurser(expr, sub_dict):
 
 def _fraction_decomp(expr):
     """Return num, den such that expr = num/den"""
-    if not isinstance(expr, Mul):
+    if not isinstance(expr, MulClass):
```

Comment:
Yes, it should all be done in a compatibility layer. Basically, for this to work, there should be no code anywhere in SymPy that does

```py
if HAS_SYMENGINE:
    ...
else:
    ...
```

That also means that the SymEngine core has to have all the same methods as SymPy (even if some of them just dispatch to SymPy Python code). For instance, if the SymEngine core doesn't have the `diff` method, then any code that takes a derivative in SymPy will have to use the above `if HAS_SYMENGINE` to do something different for the SymPy core and the SymEngine core, OR, nowhere in SymPy can use the diff method, only the function, meaning that the diff method is basically dropped from the API (and what if a user wants to use `a.diff(x)` in their code?).  

It really needs to "just work". Instead of failing with AttributeError or whatever when something isn't implemented, it should fallback to SymPy Python code. 

Regarding the speed, I would suggest just writing the simplest thing, and seeing how it affects the speed. For instance, I'm assuming you are worried that having a metaclass will slow things down. If it does, maybe there are ways to speed it up (like writing the compatibility layer in C).